hack/jenkins/e2e.sh

```diff
@@ -39,9 +39,62 @@ function join_regex_no_empty() {
     fi
 }
 
+# Print an error message and exit with status 1.
+function error_exit() {
+  echo "$1" 1>&2
+  exit 1
+}
+
+# Retry a command up to a certain numbers.
+#
+# The exit code of this function will be 0 if the command succeeds within the
+# maximum number of retries, or the exit code of the last trial of the command
+# otherwise.
+#
+# The console output of the function will be the stdout of the last and final
+# successful trial of the command. This is useful as it is common in shell
+# scripts to use stdout as a way of "returning" values from a function, and we
+# don't want error messages from earlier trials to be included in the final
+# output.
+#
+# Usage: retry <max_retries> <cmd> [cmd_param1 [...]]
+#
+# For example,
+#   #!/bin/bash
+#   ...
+#   local bar_content=""
+#   bar_content="$(retry 3 gsutil cat gs://foo-bucket/bar)"
+#   echo "${bar_content}"
+function retry() {
+  local max_retries=$1
+  local cmd="${@:2}"
+
+  local n=1
+  local stdout=""
+  local exit_code="0"
+
+  until [[ $n -gt $max_retries ]]; do
+    stdout="$(${cmd})"
```

Comment:
You are right. It wouldn't work for all cases. Instead of trying to messing with it even more, I used simple for loops wherever needed as you suggested. PTAL.